### PR TITLE
[persist] Drop the isolated runtime on its own thread

### PR DESCRIPTION
**Pull request by bkirwi**
Sqllogictest shuts down the main tokio runtime, and expects that to also stop all background tasks. However, Persist's own runtime does not shut down synchronously, since it's against the rules to drop a runtime within another runtime... which can cause background tasks to continue to run while eg. filesystem state is deleted. This slightly horrible commit forks off another thread just so we can drop the runtime synchronously and outside the runtime context, but still allow the main runtime to wait on it in a task (so that shutting down the outer runtime will also block on the inner shutdown).

### Motivation

A workaround for #20677.

### Tips for reviewer

I don't know if we actually want to do this! But if we do I think this is in a mergeable state.

I've run the sqllogictests a bunch of times on this branch and not hit any of the CI-type failures. Hard to rule out entirely, though.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [ ] If this PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way), then it is tagged with a `T-proto` label.
- [ ] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  -

### CODE REVIEW

**Review by danhhz — COMMENTED** [review]
didn't CpuHeavyRuntime get invented because blocking tasks in the main runtime were causing us issues on shutdown (probably in sqllogictest)? is this approach okay because the blocking task doesn't do anything with tokio things (timers etc)?

**Review by danhhz — COMMENTED** [review]
i'm okay with this modulo the one comment i've left. but tbh, i can't claim i 100% understood the original issue that led to CpuHeavyRuntime, so i'd be a bit uncomfortable stamping it. maybe we can get nikhil or gus to think about this and sign off on it?

File: src/persist-client/src/async_runtime.rs [hunk 1]
```diff
@@ -64,9 +66,11 @@ impl Drop for IsolatedRuntime {
...
-        self.inner
-            .take()
-            .expect("cannot drop twice")
-            .shutdown_background()
+        // However, we would like the shutdown of the outer / main runtime to block
+        // on the isolated runtime shutting down, to avoid leaking tasks. Tokio makes
+        // this very annoying! Thankfully, it's only ever relevant in tests.
+        let inner_runtime = self.inner.take().expect("cannot drop twice");
+        let thread_handle = thread::spawn(move || drop(inner_runtime));
+        task::spawn_blocking(|| "drop_isolated", || async move { thread_handle.join() });
```
[the remark this answers is not in the document]
Comment (benesch) [reply src/persist-client/src/async_runtime.rs:74]: Could stick something like this before the `spawn_blocking`?

```rust
info!("waiting for isolated runtime to drop");
let _guard = task::spawn("drop_isolated_warn", || loop {
    tokio::time::sleep(Duration::from_secs(5)).await;
    info!("still waiting for isolated runtime to drop");
}).abort_on_drop();
```

**Review by benesch — COMMENTED** [review]
This makes sense to me, and would be my recommendation for the fix if `persist` is on board with taking on this complexity.

Note that there are other tests that run in-process: the Rust tests in src/environmentd/tests. Those are a bit less susceptible to issues like this, because they create independent Cockroach databases and temporary directories, but I'll still sleep better at night knowing that when each test drops its persist clients, that actually ensures that any background work has come to a halt.

**Comment by bkirwi** [comment]
For those following along at home, we discussed elsewhere and decided that it was preferable to put the workaround in SLT instead of Persist, implemented and merged as #21028.

We may one day want to support graceful shutdown etc. from Persist, but a properly designed version will likely look pretty different...

**Pull request closed without merging**